keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1107 @@
+        // +optional
+        Env []ExecEnvVar
+ 
++       // TokenAttributes is the configuration for the service account token that will be passed to the plugin.
++       // The credential provider opts in to using service account tokens for image pull by setting this field.
++       // When this field is set, kubelet will generate a service account token bound to the pod for which the
++       // image is being pulled and pass to the plugin as part of CredentialProviderRequest along with other
++       // attributes required by the plugin.
++       //
++       // The service account metadata and token attributes will be used as a dimension to cache
++       // the credentials in kubelet. The cache key is generated by combining the service account metadata
++       // (namespace, name, UID, and annotations map with keys defined in ServiceAccountTokenAttribute.serviceAccountAnnotationKeys).
++       TokenAttributes *ServiceAccountTokenAttributes
++}
++
++// ServiceAccountTokenAttributes is the configuration for the service account token that will be passed to the plugin.
++type ServiceAccountTokenAttributes struct {
++       // ServiceAccountTokenAudience is the intended audience for the projected service account token.
++       // +required
++       ServiceAccountTokenAudience string
++
++       // ServiceAccountAnnotationKeys is the list of annotation keys that the plugin is interested in.
++       // The keys defined in this list will be extracted from the corresponding service account and passed
++       // to the plugin as part of the CredentialProviderRequest. The plugin is responsible for validating
++       // the existence of annotations and their values.
++       // This field is optional and may be empty. Plugins may use this field to extract
++       // additional information required to fetch credentials.
+        // +optional
++       ServiceAccountAnnotationKeys []string
+ }
+
+```
+
+Example credential provider configuration:
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1
+kind: CredentialProviderConfig
+providers:
+  - name: acr-credential-provider
+    matchImages:
+      - "*.registry.io/*"
+    defaultCacheDuration: "10m"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+    tokenAttributes:
+      serviceAccountTokenAudience: my-audience
+      # list of annotations that the plugin is interested in
+      # the annotation key and corresponding value in the service account will be passed to the plugin
+      serviceAccountAnnotationKeys:
+      - domain.io/identity-id
+      - domain.io/identity-type
+```
+
+##### Kubernetes API Server (KAS) Changes
+
+Add a new flag `--allowed-kubelet-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.
+
+The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)
```

Comment:
Mention that today we only verify the SA and allow all audiences.  We want to verify the audiences as well, so we need to be explicit about them here because the other sources of audiences can be observed via the kube API, but this one cannot.